### ISSUE TITLE
Fine-tuning LayoutLMv2 on multi-gpus

### ISSUE THREAD

**Issue by tiennvcs** [NielsRogge/Transformers-Tutorials#30]
Based on providing notebook at [here](https://colab.research.google.com/github/NielsRogge/Transformers-Tutorials/blob/master/LayoutLMv2/DocVQA/Fine_tuning_LayoutLMv2ForQuestionAnswering_on_DocVQA.ipynb), I reproduce LayoutLMv2 at my [repo](https://github.com/tiennvcs/docvqa/tree/main/libs/layoutlmv2). Everything is okay, but I have a trouble when fine-tuning. The fine-tuning on a large dataset like DocVQA is too long when using a single GPU. So I have followed the Pytorch instruction at [here](https://pytorch.org/tutorials/beginner/blitz/data_parallel_tutorial.html) but it generates errors.

Running command:
```bash
CUDA_VISIBLE_DEVICES=1,2 python train.py --train_config default_config --work_dir runs/train/experiment1/
```
Output:
*RuntimeError: Expected tensor for argument #1 'input' to have the same device as tensor for argument #2 'weight'; but device 1 does not equal 0 (while checking arguments for cudnn_convolution)*.

I have found this issue in some forums, but actually, it doesn't work for me.

Can you help me show me how to fine-tuning on multi-GPUs?

**Comment by tiennvcs** [comment]
Fine-tuning file can be found at [here](https://github.com/tiennvcs/docvqa/blob/main/libs/layoutlmv2/train.py). This is `train()` function.

```python
def train(model, train_data, val_data, 
        epochs, optimizer, lr, loss_log, save_freq,
        eval_freq, work_dir, logger):


    optimizer = optimizer(model.parameters(), lr=lr)

    # Measure GPU memory of model
    GPU_usage_before = get_gpu_memory_map()

    device = torch.device('cuda:0' if torch.cuda.is_available() else "cpu")
    
    if torch.cuda.device_count() > 1:
        print()
        logger.info("Let's use {} GPUs!".format(torch.cuda.device_count()))
        model = nn.DataParallel(model).cuda()
    model.to(device)
    
    gpus_usage = np.sum(get_gpu_memory_map() - GPU_usage_before)
    logger.info("GPUs usages for model: {} Mb".format(gpus_usage))

    model.train()

    min_valid_loss = np.inf
    idx = 1

    for epoch in range(1, epochs):

        logger.info("Epoch {}/{}".format(epoch, epochs))
        
        train_loss = 0.0
        for _, train_batch in enumerate(train_data):
            input_ids         = train_batch["input_ids"].to(device)
            attention_mask    = train_batch["attention_mask"].to(device)
            token_type_ids    = train_batch["token_type_ids"].to(device)
            bbox              = train_batch["bbox"].to(device)
            image             = train_batch["image"].to(device)
            start_positions   = train_batch["start_positions"].to(device)
            end_positions     = train_batch["end_positions"].to(device)

            # DEBUG for multi-gpus training
            print("Model device: {}".format(model.device_ids))
            print("Weights device: {}".format(next(model.parameters()).device))
            print("input_ids device: {}".format(input_ids.device))
            print("attention_mask device: {}".format(attention_mask.device))
            print("token_type_ids device: {}".format(token_type_ids.device))
            print("bbox device: {}".format(bbox.device))
            print("image device: {}".format(image.device))
            print("start_positions device: {}".format(start_positions.device))
            print("end_positions device: {}".format(end_positions.device))
            input()

            # zero the parameter gradients
            optimizer.zero_grad()

            # forward + backward + optimize
            outputs = model(input_ids=input_ids, attention_mask=attention_mask, token_type_ids=token_type_ids,
                            bbox=bbox, image=image, start_positions=start_positions, end_positions=end_positions)
    
            loss = outputs.loss
            
            loss.backward()
            
            optimizer.step()

            train_loss += loss.item()


            # Evaluate current model on entire validation dataset after each `eval_freq` iterations
            if idx % eval_freq == 1:
                val_loss = 0.0
                model.eval()
                for _, val_batch in enumerate(val_data):
                    
                    val_batch               = val_batch.to(device)
                    input_ids               = val_batch["input_ids"].to(device)
                    attention_mask          = val_batch["attention_mask"].to(device)
                    token_type_ids          = val_batch["token_type_ids"].to(device)
                    bbox                    = val_batch["bbox"].to(device)
                    image                   = val_batch["image"].to(device)
                    start_positions         = val_batch["start_positions"].to(device)
                    end_positions           = val_batch["end_positions"].to(device)
                    
                    outputs = model(input_ids=input_ids, attention_mask=attention_mask, token_type_ids=token_type_ids,
                                bbox=bbox, image=image, start_positions=start_positions, end_positions=end_positions)
                    loss = outputs.loss
                    # Calculate Loss
                    val_loss += loss.item()
                
                logger.info("Iterations: {:<6} - epoch: {:<3} - train_loss: {:<6} - val_loss: {:<6}".format(idx, epoch, train_loss/eval_freq, val_loss/len(val_data)))
                loss_log.info("Iterations: {:<6} - epoch: {:<3} - train_loss: {:<6} - val_loss: {:<6}".format(idx, epoch, train_loss/eval_freq, val_loss/len(val_data)))

                    
                if min_valid_loss > val_loss/len(val_data):
                    logger.info("Found best model !! Validation loss descreased from {} to {}".format(min_valid_loss, val_loss/len(val_data)))
                    torch.save(model.state_dict(), os.path.join(work_dir, 'best'+'.pth'))
                    min_valid_loss = val_loss/len(val_data)

                # Save model each save_freq iteration
                if idx % save_freq == 1:
                    logger.info("Saving model to {}".format(os.path.join(work_dir, str(idx).zfill(5)+'.pth')))
                    torch.save(model.state_dict(), os.path.join(work_dir, str(idx).zfill(5)+'.pth'))

                # Reset training loss
                train_loss = 0.0
                
            idx += 1


    logger.info("Done !")
    logger.info("The minimum on validation {}".format(min_valid_loss))

    return model
```

**Comment by NielsRogge** [comment]
Hi,

Thanks for your interest in using LayoutLMv2. You should probably change this line: `model = nn.DataParallel(model).cuda()` to just this: `model = nn.DataParallel(model)`.

**Comment by LOC333** [comment]
I am having the same problem. Training on a single GPU works fine, when I try to use DataParallel to train on multi GPU it throws an error
"RuntimeError: Expected tensor for argument #1 'input' to have the same device as tensor for argument #2 'weight'; but device 1 does not equal 0 (while checking arguments for cudnn_convolution". 

I have followed the DataParallel tutorial

`device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
model = LayoutLMv2ForSequenceClassification.from_pretrained("layoutlmv2-base-uncased", num_labels=len(set(labels)))
if torch.cuda.device_count() > 1:
    print("Let's use", torch.cuda.device_count(), "GPUs!")
    model = nn.DataParallel(model)

model.to(device)`

My inputs are all on cuda:0 also. 

`torch.cuda.device_count()` returns 4

**Comment by NielsRogge** [comment]
As explained in the [docs](https://huggingface.co/transformers/model_doc/layoutlmv2.html) of LayoutLMv2:

> If you want to train the model in a distributed environment, make sure to call `synchronize_batch_norm()` on the model in order to properly synchronize the batch normalization layers of the visual backbone.

Apologies if this was a bit hidden, but I removed the distributed training code from the modeling file to make it more clean. To train in a distributed environment, you first need to do `model.synchronize_batch_norm()` before calling `model = nn.DataParallel(model)`.

**Comment by tiennvcs** [comment]
Thank you for

> As explained in the [docs](https://huggingface.co/transformers/model_doc/layoutlmv2.html) of LayoutLMv2:
> 
> > If you want to train the model in a distributed environment, make sure to call `synchronize_batch_norm()` on the model in order to properly synchronize the batch normalization layers of the visual backbone.
> 
> Apologies if this was a bit hidden, but I removed the distributed training code from the modeling file to make it more clean. To train in a distributed environment, you first need to do `model.synchronize_batch_norm()` before calling `model = nn.DataParallel(model)`.

Yeah... It's really helpful for me. But I got a problem with synchronize_batch_norm(), the problem said that `LayoutLMv2ForQuestionAnswering` object hasn't attribute `synchronize_batch_norm`. Could you help solve this issue?

```cmd
Traceback (most recent call last):
  File "train.py", line 189, in <module>
    main(args)
  File "train.py", line 165, in main
    model = AutoModelForQuestionAnswering.from_pretrained(MODEL_CHECKPOINT).synchronize_batch_norm()
  File "/home/tiennv/anaconda3_v2/envs/transformer_env/lib/python3.7/site-packages/torch/nn/modules/module.py", line 948, in __getattr__
    type(self).__name__, name))
AttributeError: 'LayoutLMv2ForQuestionAnswering' object has no attribute 'synchronize_batch_norm'
```

**Comment by NielsRogge** [comment]
Yes, you should call it on the base model, i.e. `model.layoutlmv2.synchronize_batch_norm()`.

The method itself is defined for the base model.

**Comment by LOC333** [comment]
> ```batchfile
> Traceback (most recent call last):
> ```

I think it may be `model.layoutlmv2.visual.synchronize_batch_norm()`

**Comment by tiennvcs** [comment]
It raises the error again.

```cmd
Traceback (most recent call last):
  File "train.py", line 189, in <module>
    main(args)
  File "train.py", line 170, in main
    work_dir=args['work_dir'], logger=logger, eval_freq=eval_freq, gpu_ids=gpu_ids)
  File "train.py", line 33, in train
    model.layoutlmv2.visual.synchronize_batch_norm()
  File "/home/tiennv/anaconda3_v2/envs/transformer_env/lib/python3.7/site-packages/transformers/models/layoutlmv2/modeling_layoutlmv2.py", line 594, in synchronize_batch_norm
    raise RuntimeError("Make sure torch.distributed is set up properly.")
RuntimeError: Make sure torch.distributed is set up properly.
```

**Comment by NielsRogge** [comment]
You probably first need to call `torch.distributed.init_process_group()` before starting training.

**Comment by LOC333** [comment]
Thank you for your quick response @NielsRogge. Do we need to use DistributedDataParallel or will DataParallel work? If you were able to share the code that you used for distributed training that would be very helpful.

**Comment by tiennvcs** [comment]
I reproduce the training code from **DataParallel** to **DistributedDataParallel**, It does not release bugs in training, **but it does not print any log or running**.
**Could show me what is wrong with my code?**

This is my code.
```python
import torch
import argparse
import os, time
import json, tqdm
from utils import find_highest_score_answer, load_feature_from_file, create_logger
from transformers import AutoModelForQuestionAnswering
from transformers import AutoTokenizer


def inference(model, data, tokenizer):


    device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
    model.eval()

    results = []

    for _, batch in tqdm.tqdm(enumerate(data)):
        
        input_ids               = batch["input_ids"].to(device)
        attention_mask          = batch["attention_mask"].to(device)
        token_type_ids          = batch["token_type_ids"].to(device)
        bbox                    = batch["bbox"].to(device)
        image                   = batch["image"].to(device)
        
        outputs = model(input_ids=input_ids, attention_mask=attention_mask, token_type_ids=token_type_ids,
                    bbox=bbox, image=image)
        start_logits = outputs.start_logits.detach().cpu().numpy()
        end_logits   = outputs.end_logits.detach().cpu().numpy()
        start_indices, end_indices = find_highest_score_answer(
                                    start_scores=start_logits, end_scores=end_logits)
        
        input_ids    = input_ids.cpu().numpy()
        question_ids = batch["question_id"].detach().cpu().numpy().tolist()

        for question_id, input_id, s, e in zip(question_ids, input_ids, start_indices, end_indices):
            predicted_answer = tokenizer.decode(input_id[s:e+1])
            decoding_string  = tokenizer.decode(input_id)
            question         = decoding_string[decoding_string.find('[CLS]')+5:decoding_string.find('[SEP]')]
            results.append({
                "questionId": question_id,
                "question": question,
                "answer": predicted_answer,
            })
    return results


def main(args):

    output_dir = os.path.join(args['output_dir'], args['weights'].split("/")[-2])
    if not os.path.exists(output_dir):
        try:
            print("Creating {} directory".format(output_dir))
            os.mkdir(output_dir)
        except:
            print("INVALID OUTPUT DIRECTORY")
            exit(0)
    
    logger = create_logger(file_path=os.path.join(output_dir, 'inference.log'))

    # Load dataset
    logger.info("Loading dataset from {} ...".format(args['input_dir']))
    eval_data = load_feature_from_file(path=args['input_dir'], batch_size=2)
    logger.info("The number of sample is {} ...".format(len(eval_data.dataset)))

    # Load model
    logger.info("Loading model from {} ...".format(args['weights']))
    model = AutoModelForQuestionAnswering.from_pretrained(args['model']).cuda()
    # model.load_state_dict(torch.load(args['weights'], map_location='cuda'))
    
    # Load tokenizer
    tokenizer = AutoTokenizer.from_pretrained(args['model'])

    # Inference
    logger.info("Start inference ...")
    start_time  = time.time()
    results     = inference(model=model, data=eval_data, tokenizer=tokenizer)
    end_time    = time.time()
    logger.info("Total inference time {} seconds".format(end_time-start_time))
    
    # save inference results to disk
    output_file = os.path.join(output_dir, args['input_dir'].split("/")[-2] + '.json')
    with open(output_file, 'w') as f:
        json.dump(results, f, indent=4)
    logger.info("DONE ! Check the inference results at {}".format(output_file))



if __name__ == '__main__':
    
    parser = argparse.ArgumentParser(description='Inference on DocVQA dataset.')

    parser.add_argument('--input_dir', required=True,
        help='The input feature extracted from DocVQA dataset. I can be train/val/test subset.' ,
    )

    parser.add_argument('--model', default='microsoft/layoutlmv2-base-uncased',
        help='The model architecture.'
    )
    parser.add_argument('--weights', required=True,
        help='The path to model weights',
    )

    parser.add_argument('--output_dir', required=True,
        help='The output directory'
    )

    args = vars(parser.parse_args())

    main(args)
```

The output in terminal like below.
```bash
(transformer_env) root@ae94a4e6c92d:/mlcv/WorkingSpace/NCKH/tiennv/vqa_thesis/docvqa/libs/layoutlmv2# CUDA_VISIBLE_DEVICES=1,2 python train.py --work_dir ./runs/train/test_multi-gpus --train_config default_config
2021-09-26 10:11:49,801 - INFO - Loading training configuration ...
2021-09-26 10:11:49,802 - INFO - Configuration: {'optimizer': <class 'torch.optim.adam.Adam'>, 'lr': 0.0001, 'epochs': 2, 'batch_size': 2, 'momentum': 0.9, 'eval_freq': 1, 'save_freq': 1, 'num_workers': 4}
2021-09-26 10:11:49,803 - INFO - Loading training dataset from /mlcv/Databases/DocVQA_2020-21/task_1/extracted_features/layoutlmv2/train ...
2021-09-26 10:11:49,953 - INFO - Loading validation dataset from /mlcv/Databases/DocVQA_2020-21/task_1/extracted_features/layoutlmv2/val ...
2021-09-26 10:11:49,977 - INFO - Training size: 39456 - Validation size: 5344
2021-09-26 10:11:49,978 - INFO - Loading pre-training model from microsoft/layoutlmv2-base-uncased checkpoint
Some weights of the model checkpoint at microsoft/layoutlmv2-base-uncased were not used when initializing LayoutLMv2ForQuestionAnswering: ['layoutlmv2.visual.backbone.bottom_up.res4.0.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.7.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.11.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.12.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.5.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.19.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.1.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.15.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res3.1.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.19.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res2.0.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res5.0.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.21.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res5.2.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.16.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.21.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.14.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.2.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.8.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res3.2.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.9.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.4.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res2.2.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.22.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res5.0.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.5.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res3.3.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res2.2.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.4.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.17.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res3.1.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.11.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.1.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.3.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.7.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.8.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res3.3.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.13.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.22.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res5.2.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res5.0.shortcut.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res2.2.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.6.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res5.1.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res3.2.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.13.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res2.0.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res2.0.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.5.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.20.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.16.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.0.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.7.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.14.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.11.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res2.1.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res3.3.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.2.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.3.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.12.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.20.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res5.0.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.3.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.14.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.10.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res3.0.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.18.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.19.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.21.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.stem.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.17.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.0.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.12.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.1.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res5.2.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res3.0.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.8.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res2.1.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.0.shortcut.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.22.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res3.1.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res3.2.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.15.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res3.0.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res5.1.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res2.1.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.9.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.18.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.10.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.20.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.17.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.18.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.10.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.6.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res3.0.shortcut.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.6.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.16.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.2.conv1.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res5.1.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.4.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.9.conv3.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.13.conv2.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res2.0.shortcut.norm.num_batches_tracked', 'layoutlmv2.visual.backbone.bottom_up.res4.15.conv3.norm.num_batches_tracked']
- This IS expected if you are initializing LayoutLMv2ForQuestionAnswering from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing LayoutLMv2ForQuestionAnswering from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of LayoutLMv2ForQuestionAnswering were not initialized from the model checkpoint at microsoft/layoutlmv2-base-uncased and are newly initialized: ['qa_outputs.weight', 'qa_outputs.bias', 'layoutlmv2.visual_segment_embedding']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Running DDP with model parallel example on cuda:0 device
Running DDP with model parallel example on cuda:1 device
GPUs usages for model: 6721 Mb
Epoch 1/2
GPUs usages for model: 6721 Mb
Epoch 1/2
```

**Comment by athewsey** [comment]
Sorry @NielsRogge / all, maybe this is a silly question but:

When using the Trainer API instead of manually setting up torch DP and the training loop... where could `model.layoutlmv2.synchronize_batch_norm()` be called? Maybe there's a callback that could be hooked into after Trainer sets up the distribution but before the loop starts?

(I'm running in SageMaker and trying to keep to Trainer to stay agnostic to whether the user is using SMDistributed or not)
